[SPARK-31703][SQL] Parquet RLE float/double are read incorrectly on big endian platforms

tinhto-000 opened the pull request:
### What changes were proposed in this pull request?
This PR fixes the issue introduced during SPARK-26985.

SPARK-26985 changes the `putDoubles()` and `putFloats()` methods to respect the platform's endian-ness.  However, that causes the RLE paths in VectorizedRleValuesReader.java to read the RLE entries in parquet as BIG_ENDIAN on big endian platforms (i.e., as is), even though parquet data is always in little endian format.

The comments in `WriteableColumnVector.java` say those methods are used for "ieee formatted doubles in platform native endian" (or floats), but since the data in parquet is always in little endian format, use of those methods appears to be inappropriate.

To demonstrate the problem with spark-shell:

```scala
import org.apache.spark._
import org.apache.spark.sql._
import org.apache.spark.sql.types._
    
var data = Seq(
  (1.0, 0.1),
  (2.0, 0.2),
  (0.3, 3.0),
  (4.0, 4.0),
  (5.0, 5.0))
  
var df = spark.createDataFrame(data).write.mode(SaveMode.Overwrite).parquet("/tmp/data.parquet2")
var df2 = spark.read.parquet("/tmp/data.parquet2")
df2.show()
```

result:

```scala
+--------------------+--------------------+                                     
|                  _1|                  _2|
+--------------------+--------------------+
|           3.16E-322|-1.54234871366845...|
|         2.0553E-320|         2.0553E-320|
|          2.561E-320|          2.561E-320|
|4.66726145843124E-62|         1.0435E-320|
|        3.03865E-319|-1.54234871366757...|
+--------------------+--------------------+
```

Also tests in ParquetIOSuite that involve float/double data would fail, e.g., 

- basic data types (without binary)
- read raw Parquet file

/examples/src/main/python/mllib/isotonic_regression_example.py would fail as well.

Purposed code change is to add `putDoublesLittleEndian()` and `putFloatsLittleEndian()` methods for parquet to invoke, just like the existing `putIntsLittleEndian()` and `putLongsLittleEndian()`.  On little endian platforms they would call `putDoubles()` and `putFloats()`, on big endian they would read the entries as little endian like pre-SPARK-26985.

No new unit-test is introduced as the existing ones are actually sufficient.

### Why are the changes needed?
RLE float/double data in parquet files will not be read back correctly on big endian platforms.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
All unit tests (mvn test) were ran and OK.

tinhto-000 committed:
SPARK-31703

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

tinhto-000 committed:
missing ;

maropu commented:
ok to test

SparkQA commented:
**[Test build #127203 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127203/testReport)** for PR 29383 at commit [`2261c09`](https://github.com/apache/spark/commit/2261c09e5fb6090372022517de67ad02f2c3dad6).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31824/
Test PASSed.

maropu commented:
cc: @rednaxelafx @kiszk @srowen

maropu commented:
also cc @ScrapCodes and @zhengruifeng because this bug is realted to branch-3.0/branch-2.4 and they are the release managers.

dongjoon-hyun commented:
cc @vibhutisawant since he is the original author.

srowen commented:
Seems reasonable - is there any case where we would need putFloats separately from putFloatsLittleEndian? like would you ever legitimately not care about endianness of the serialization? (may be a dumb question)

srowen commented on sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/OffHeapColumnVector.java:
```diff
@@ -425,6 +425,19 @@ public void putFloats(int rowId, int count, byte[] src, int srcIndex) {
 
+  @Override
+  public void putFloatsLittleEndian(int rowId, int count, byte[] src, int srcIndex) {
+    if (!bigEndianPlatform) {
```
Really doesn't matter, but you could flip the comparison here to avoid the negation

tinhto-000 replied:
agreed, it definitely will look better.  will change in next commit.  thanks!

tinhto-000 replied:
Actually now I look at the existing `putDoublesLittleEndian()` and `putFloatsLittleEndian()` methods and they do `if (!bigEndianPlatform)` as well (and thus how I copied and pasted from those).

@srowen Do you think we can leave the negation alone for uniformity, or go ahead and change?

srowen replied:
Just leave it as-is for consistency, yeah. OK I see we have lots of "as little endian" variants so maybe this is the best approach for symmetry.

I dont' really know enough about this part to validate the change but it seems plausible.

SparkQA commented:
**[Test build #127203 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127203/testReport)** for PR 29383 at commit [`2261c09`](https://github.com/apache/spark/commit/2261c09e5fb6090372022517de67ad02f2c3dad6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127203/
Test PASSed.

tinhto-000 commented:
> Seems reasonable - is there any case where we would need putFloats separately from putFloatsLittleEndian? like would you ever legitimately not care about endianness of the serialization? (may be a dumb question)

It's a good question.  It *seems* like column compression drives putFloats() and I'm *guessing* that's the reason for the original SPARK-26985.

dongjoon-hyun commented:
cc @ScrapCodes since this is a correctness issue.

dongjoon-hyun commented:
I raised the priority of SPARK-31703 to `Blocker` with `Target Version` 2.4.7 and 3.0.1.

kiszk commented:
Good catch. This change looks good.   
Could you please add test cases to [ColumnarBatchSuite.scala](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala) like `putIntsLittleEndian()`?

HyukjinKwon reviewed (commented):
The change looks good to me too. +1 for adding a test.

cloud-fan commented:
LGTM. +1 for adding tests

maropu commented:
Looks fine, too.

tinhto-000 committed:
SPARK-31703 add tests

SparkQA commented:
**[Test build #127336 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127336/testReport)** for PR 29383 at commit [`f160da0`](https://github.com/apache/spark/commit/f160da0035d8b1df08b75832f5871f7afc5cb839).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31955/
Test PASSed.

tinhto-000 commented:
@kiszk thanks for the great suggestion and point out.  Turned out ColumnarBatchSuite.scala failed on big endian as well.

The "Double APIs" and "Float APIs" are using `putFloats()` and `putDoubles()` on little endian stream they explicitly created , and those failed on big endian.

I modified those 2 tests and added 2 [SPARK-31703] tests, and the tests look fine on both x86 and z/OS (s390) now.

SparkQA commented:
**[Test build #127336 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127336/testReport)** for PR 29383 at commit [`f160da0`](https://github.com/apache/spark/commit/f160da0035d8b1df08b75832f5871f7afc5cb839).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127336/
Test PASSed.

cloud-fan commented:
thanks, merging to master/3.0!

kiszk commented on sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala:
```diff
@@ -447,13 +447,6 @@ class ColumnarBatchSuite extends SparkFunSuite {
       Platform.putFloat(buffer, Platform.BYTE_ARRAY_OFFSET, 2.234f)
       Platform.putFloat(buffer, Platform.BYTE_ARRAY_OFFSET + 4, 1.123f)
 
-      if (ByteOrder.nativeOrder().equals(ByteOrder.BIG_ENDIAN)) {
```
Thank you. This code did not make sense due to [the old implementation](https://github.com/tinhto-000/spark/commit/8a1ce4899fb9f751dedaaa34ea654dfbc8330852)
